Added support of quoted strings in Timidity config

Dasperal opened the pull request:
Resolves #334

An argument of property in Timidity config can be surrounded with `"` or `'` characters to prevent splitting into several arguments if it contains spaces or other delimiter characters.

If an argument itself contains `"` characters it should be surrounded with `'` characters and vice versa.

sezero commented:
Does this match Timidity++?

Dasperal commented:
It is based on Timidity++ but matches the original SDL_Mixer's Timidity behavior except for quoted strings. And it appears that space between a quoted string and the next argument became optional.

Do I need to make config file parsing fully Timidity++ compatible?

sezero commented:
> It is based on Timidity++ but matches the original SDL_Mixer's Timidity behavior except for quoted strings.

OK, 

> And it appears that space between a quoted string and the next argument became optional.

Eh? Say, `0 "instruments/acpiano"amp=125` is valid now?

> Do I need to make config file parsing fully Timidity++ compatible?

No, only and only quoted string _so that_ paths with spaces would be handled correctly.
And please test carefully - e.g. if we hit a `"` do we require it closing match, etc?

Dasperal commented:
Updated patch. Fixed bug which stops processing of the line after an unquoted argument. Didn't notice it originally because I had a Timidity config with properties with only one argument.

Tested following scenarios of splitting line by words:
1) line `dir abc def ghi jkl`
will be split to `dir`, `abc`, `def`, `ghi`, `jkl`
2) line `dir "abc def" ghi jkl`
will be split to `dir`, `abc def`, `ghi`, `jkl`
3) line `dir "abc"def "ghi"jkl`
will be split to `dir`, `abc`, `def`, `ghi`, `jkl`
4) line `dir "abc""def""ghi""jkl"`
will be split to `dir`, `abc`, `def`, `ghi`, `jkl`
5) line `dir abc"def" ghi"jkl"`
will be split to `dir`, `abc"def"`, `ghi"jkl"`
6) line `dir "abc def ghi jkl`
will be split to `dir`, `"abc`, `def`, `ghi`, `jkl`
7) line `dir "abc def ghi' jkl`
will be split to `dir`, `"abc`, `def`, `ghi'`, `jkl`
8) line `dir 'abc "def" ghi' jkl`
will be split to `dir`, `abc "def" ghi`, `jkl`

I think scenarios 3, 4, 6, 7 should be errors.

sezero commented:
> I think scenarios 3, 4, 6, 7 should be errors.

Indeed.

Case 8 is weird, though. Anyone does that?

sezero commented:
My point is, only the first scenario is sensible (unless I'm proven otherwise),
e.g. path arguments to `dir`.  So we should do something like, in pseudo-code:
```c
  seenquote  = 0;
  [hit 1st quote]
  if (!isspace(*beginquote)) {
    error; /* only arguments can be quoted */
  }
  seenquote++;
  [find endquote]
  if (!endquote) {
    error;
  }
  [then :]
  if (seenquote > 2) {
    error; /* multiple quoted args: shouldn't be possible, yes? */
  }
```

Dasperal commented:
>Why do you want to accept case 5??

In the original implementation with "SDL_strtokr", the resulting words may contain any garbage characters. In this particular case, it may look like "You wanted a quoted argument but forgot a space" so I can make this an error.

>Case 8 is weird, though. Anyone does that?

I don't know why it can be needed nor if it ever will.

Errors are just
``` C
SNDDBG(("%s: line %d: Error message", name, line));
goto fail;
```
right?

sezero commented:
> > Why do you want to accept case 5??
> 
> In the original implementation with "SDL_strtokr", the resulting words may contain any garbage characters. In this particular case, it may look like "You wanted a quoted argument but forgot a space" so I can make this an error.

Let's make that one an error

> > Case 8 is weird, though. Anyone does that?
> 
> I don't know why it can be needed nor if it ever will.

Indeed, let's make that an error too

> Errors are just
> ```c
> SNDDBG(("%s: line %d: Error message", name, line));
> goto fail;
> ```
> right?

Yes.

Dasperal commented:
No response to the last push so far, so I think I should leave a comment.

Updated patch. Mentioned cases (3 - 8) are treated as errors now.

Is there anything else I need to do or test?

sezero commented:
Sorry, was a little busy + lazy + etc

Will test soon and if no issues, will merge.  Would you agree that I apply this to [libtimidity](https://github.com/sezero/libtimidity) and [SDL_sound](https://github.com/icculus/SDL_sound) too?

Dasperal commented:
> Would you agree that I apply this to [libtimidity](https://github.com/sezero/libtimidity) and [SDL_sound](https://github.com/icculus/SDL_sound) too?

Yes

sezero commented:
@icculus, @slouken: Does this patchset look good to you? Do you see any gotchas?

slouken commented:
Yep, looks good, thanks!

sezero commented:
Patch now applied to SDL_sound.

sezero commented:
Patch now applied to libtimidity too.

sezero commented:
@Dasperal, @slouken: I pushed https://github.com/libsdl-org/SDL_mixer/commit/d1cff77caaf84783b510d603e466b566852a4e94 for better readability: please make sure that I didn't screw-up.

Dasperal commented:
Tested. Quoted strings still work
